File: temperature-analytics-api/src/main/java/io/kontakt/apps/rest/AnomalyController.java
package io.kontakt.apps.rest;

import io.kontak.apps.Filter;
import io.kontak.apps.usecase.FindAnomalies;
import io.kontakt.apps.rest.dto.AnomalyResponse;
import io.kontakt.apps.rest.dto.AnomalyThermometersResponse;
import io.kontakt.apps.rest.mapper.AnomalyDtoMapper;
import io.kontakt.apps.rest.mapper.AnomalyThermometerIdDtoMapper;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.math.BigDecimal;
import java.util.Optional;

@RestController
public class AnomalyController {

    private final FindAnomalies findAnomalies;

    @Value("${temperature.default.threshold}")
    private BigDecimal defaultThreshold;

    public AnomalyController(FindAnomalies findAnomalies) {
        this.findAnomalies = findAnomalies;
    }

    @GetMapping("/anomalies/thermometer/{thermometerId}")
    public AnomalyResponse getAnomaliesByThermometerId(@PathVariable final String thermometerId) {
        final var anomalies = findAnomalies.all()
                .filterBy(
                        Filter.builder().thermometerId(thermometerId).build())
                .stream().
                map(AnomalyDtoMapper::toAnomalyDto).toList();

        return new AnomalyResponse(anomalies);
    }

    @GetMapping("/anomalies/room/{roomId}")
    public AnomalyResponse getAnomaliesByRoomId(@PathVariable final String roomId) {
        final var anomalies = findAnomalies.all()
                .filterBy(
                        Filter.builder().roomId(roomId).build())
                .stream().
                map(AnomalyDtoMapper::toAnomalyDto).toList();
        return new AnomalyResponse(anomalies);
    }

    // if statement is present because of requirement getting threshold from properties
    // otherwise we could use in @RequestParam param defaultValue.
    @GetMapping("/anomalies/thermometers")
    public AnomalyThermometersResponse getThermometersWithAnomalies(
            @RequestParam(name = "threshold") final Optional<BigDecimal> threshold) {
        BigDecimal thresholdToUse = defaultThreshold;

        if (threshold.isPresent()) {
            thresholdToUse = threshold.get();
        }

        final var thermometerIds = findAnomalies.all()
                .filterBy(
                        Filter.builder().threshold(thresholdToUse).build())
                .stream().
                map(AnomalyThermometerIdDtoMapper::toAnomalyThermometerIdDto).toList();

        return new AnomalyThermometersResponse(thermometerIds);
    }
}

File: temperature-analytics-api/src/main/kotlin/io/kontak/apps/temperatureanalyticsapi/controller/AnomalyController.kt
package io.kontak.apps.temperatureanalyticsapi.controller

import io.kontak.apps.event.Anomaly
import io.kontak.apps.event.Thermometer
import io.kontak.apps.temperatureanalyticsapi.service.AnomalyService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Flux

@RestController
class AnomalyController(@Autowired private val anomalyService: AnomalyService) {

    @GetMapping("/anomalies/thermometer/{thermometerId}")
    fun getAnomaliesByThermometerId(@PathVariable thermometerId: String): Flux<Anomaly> =
        anomalyService.findAnomaliesByThermometerId(thermometerId)

    @GetMapping("/anomalies/room/{roomId}")
    fun getAnomaliesByRoomId(@PathVariable roomId: String): Flux<Anomaly> =
        anomalyService.findAnomaliesByRoomId(roomId)

    @GetMapping("/thermometers/anomalies")
    fun getThermometersWithAnomaliesAboveThreshold(@RequestParam(required = false) threshold: Int?): Flux<Thermometer> =
        anomalyService.findThermometersWithAnomaliesAboveThreshold(threshold)
}
